### PR TITLE
RFC: drivers/qspi: add QSPI API

### PR DESCRIPTION
**Pull request by kamlaz**
Adds QSPI API

Signed-off-by: Kamil Lazowski <Kamil.Lazowski@nordicsemi.no>

**Clarification on PR content:**
Author intention is to introduce API for the `Serial Memory Protocol Engine` with serial back-end using 4 line SPI. The peripheral of `Serial Memory Protocol Engine` gives huge acceleration in I/O operation.
Wort to emphasize that such a peripheral not necessarily allow to RAW access to SPI bus at all.

Such peripherals are incorporated into some of nRF, NXP, STM and Microchip SoC.

### CODE REVIEW

**Review by tbursztyka — DISMISSED** [review]
Enhanced SPI modes are already discussed here  #17902 and does not require an entirely new API.

File: include/drivers/qspi.h [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+
+	/* Defines how many bits are used for address (8/16/24/32)
+	 * 0x00: 8 bit address field
+	 * 0x01: 16 bit address field
+	 * 0x02: 24 bit address field
+	 * 0x03: 32 bit address field
+	 */
+	u8_t address : 2;
+
+	/* Specifies the Chip Select High Time.
```
Comment (pabigot) [inline include/drivers/qspi.h]: If we need this, and we can't guarantee the driver implementation will satisfy timing requirements, we need a lot more detail on how it's to be represented in a generic way.  Probably nanoseconds would make sense, but again this should be coming from devicetree and probably won't be stored in a `u8_t`.

**Review by pabigot — COMMENTED** [review]
What's the goal of this driver: something specific to supporting JEDEC SPI NOR devices that are DSPI/QSPI-capable, or a generic QSPI driver similar to the existing SPI one?

The in-tree `drivers/flash/soc_flash_nios2_qspi` takes a much simpler approach of supporting the existing flash API using a vendor-specific HAL.  However, this board may be removed, so its solution may not be relevant.

This PR seems to be proposing the second approach.  But QSPI is simply an extended SPI.  The configuration properties and operations should, on the whole, be consistent with the SPI API.  In fact, it may be appropriate to simply extend the SPI API.  But this proposal bears no relation to the existing SPI driver.

There are two levels of configuration: one for the bus, providing:
* One or more chipselect signals;
* Up to four I/O signals (standard SPI, DSPI, and QSPI)
* Possibly bus frequency

and one for devices on the bus, providing:
* which chipselect is used;
* CPOL, CPHA configuration
* address length
* operating mode (1-1-1, 1-1-2, 1-1-4, 1-2-2, 1-4-4, ...)
* other device information including frequency

The existing SPI driver pre-dates devicetree so some of the client information is either hard-coded into specific drivers or passed through Kconfig options.  It should all be defined in devicetree.

We need to see what the devicetree bindings for the driver and nodes will be.

I've added a few comments below related to how this impacts the current proposal, before I saw @tbursztyka's comment about #17902 and stopped.

My initial perspective is that we should rework and extend the existing SPI API, which has been mostly ignored for well over a year and could use some updating.  At  least three vendors beyond Nordic support QSPI (atmel, nxp, stm32, possibly cypress and microchip).  We should do some coordination with all stakeholders to avoid producing another API that's only relevant to a single vendor.

**Comment by carlescufi** [comment]
@tbursztyka @kamlaz @pabigot I will add #17902 to the API rework project and we can discuss it next week there.

**Comment by kamlaz** [comment]
**MAIN GOAL**
The main goal of this PR is to provide suitable driver for multiple SPI NOR Flash devices.
From hardware point of view situation looks like on presented picture:
image

Imagine that SPI is QSPI (so there should be 4 data lines) and each slave has its own CS pin that is hardware controlled by QSPI peripherial
Main features of presented driver are:
	

1. Possibility of driving few different memories (each with different QSPI configuration)

2. Hardware controlled CS pin guarantees more efficient data flow 
	   (minimum time waste from going CS low to start transmission and at the end of the  transmission). Of course it is mandatory for peripherial to support few hardware CS pins.

3. Using QSPI driver with flash driver will enable user to create consistent file system that is splitted on few flash memories.

4. Easier and faster communication with flash memories. It is due to API is tailored to support QSPI commands with minimum overhead.

5.  QSPI controls by hardware four data lines and DDR mode - something that gives extremly boost, and is unavaiable for SPI


 **API HIERARCHY**
Hierarchy will look like this:
  
     -----------
	|  FS  API  |
	 -----------
	      |
	      |
     -----------
	| FLASH API |
	 -----------
	      |
	      |
     -----------
	| QSPI  API |
	 -----------		  

At first glance it could be a strange idea to have both "FLASH API" and "QSPI API" layers. But it is not without reason.
Imagine, that you have mounted file system that consists of two flash memories. Typical flow will look like this:
1. FS wants to read data from #ADDR. It calls flash_api_read()
2. flash_api_read() launches. As input argument is takes address(or offset, not important now) from it have to read data. 
   Flash layer contains information about all devices that were defined as flash memories. Informations are i.e storage size,
   protocol configuration (number of used data lines), etc. 
   Flash layer also contains flash map, so it knows whence to read data(from which flash memory).
   So now FLASH_API just call qspi_configure() to set configuration for Flash #1 and then calls qspi_read()
3. qspi_read() launches and it fetches data from the flash memory and notifiy flash API, that transfer has finished

Situation looks the same, when FS wants to read data located on the #2 flash memory.
In contraction : Let driver do what driver should do - driver the peripheral. 
Flash driver should take care of calling QSPI driver functions to talk with flash memories. 
So flash driver has to be "above" QSPI driver, like presented above.


**DTS**

To relize presented above functionality we need a specialy tailored DTS: 
```
&qspi {
	compatible = "nordic,nrf-qspi";
	status = "okay";
	sck-pin = <19>;
	io00-pin = <20>;
	io01-pin = <21>;
	io02-pin = <22>;
	io03-pin = <23>; 
	max-frequency = <32000000>;
	mem1: mem1@0 {
		compatible = "jedec,qspi-nor";
		reg = <0>; 
		cs-pin = <17>;
		frequency = <16000000>;
		qspi-mode = <3>;
		data-lines = <4>;
		address-size = <24>;
		cs-high-time = <1>;
		label = "MEM1";
		jedec-id = [c2 28 17];
		size = <67108864>;
		has-be32k;
		has-ddr-mode;
	};
};
```

As you can see, QSPI peripherial is defined by pins (SCK,IO00,IO01,IO02,IO03) and maximum alowable frequency.
It is absolutely minimum for QSPI driver. The rest of the required data are configured in the memory field, sice they are memory-specific.
The aim was to provide possibility of working with multiple flash devices and each has different configuration, i.e:
```	
mem1: mem1@0 {
		compatible = "jedec,qspi-nor";
		reg = <0>; 
		cs-pin = <17>;
		frequency = <16000000>;
		qspi-mode = <3>;
		data-lines = <4>;
		address-size = <24>;
		cs-high-time = <1>;
		label = "MEM1";
		jedec-id = [c2 28 17];
		size = <67108864>;
		has-be32k;
		has-ddr-mode;
	};
	
	mem2: mem2@1 {
		compatible = "jedec,qspi-nor";
		reg = <1>; 
		cs-pin = <18>;
		frequency = <8000000>;
		qspi-mode = <0>;
		data-lines = <4>;
		address-size = <32>;
		cs-high-time = <1>;
		label = "MEM1";
		jedec-id = [c2 28 17];
		size = <67108864>;
	};
```
As you can see, QSPI peripherial will work with two flash memories and each has completely different configuration, but
both memories share SCK,IO00,IO01,IO02,IO03 pins. Each memory has independent configuration.

**QSPI VS SPI**
Very important fact is that SPI is not the same as QSPI. They are similar to some point, but in general they are different.
The main difference is that QSPI is implementing protocol used to communicate with flash memories. In general it consists of:

- command
- address (if command requires it)
- payload
Example:  [CMD][ADDR][PAYLOAD]
Nordic's hardware QSPI takes care about this protocol.
I.e when you want to write some data to the flash you have to:

1. 	Check if memory is ready for transfer
2. 	Send WREN (write enable) command
3. 	Send write command with payload

	
You can do it using SPI, but you need to preapre 3 separate commands. 
In our implementation all steps are done by hardware, so you just have to tell where to write data and peripherial will do the rest. 
Huge saving of time and processing power - something that is worth investigating.

**Comment by tbursztyka** [comment]
> Very important fact is that SPI is not the same as QSPI. They are similar to some point, but in general they are different.
The main difference is that QSPI is implementing protocol used to communicate with flash memories. 

Do you have any reference to backup this, I mean in general about dual/quad/octal modes?

Because so far, nothing, nowhere, it is said to be like that. These modes are improvements to existing SPI bus. It is still a bus we are talking about, not a protocol. The fact some hw provide optimization for a protocol (here flash one) is different (see below).

What your described is exactly what has been done with drivers/flash/spi_nor.c : a flash device accessed through SPI bus. Only difference being that your use case can take advantage of dual/quad/octal modes (i.e.: more data lines for faster transaction) + your vendor specific hw optimization (see below)

To me it seems you are confusing the bus and the protocol which is running on top.
The format of cmd/address/payload was indeed formalized into these modes but just as an hw optimization. But:

- this format is not flash specific.
- nothing dictates that this format must be used within these modes. It is in fact possible to just tx/rx some data directly as usual. (no command, no address, just data)

Note that a controller that can do dual/quad/octal spi, can still do the standard mode as well.
So with you solution we would need 2 drivers: one SPI, one QSPI (knowing that 95% of the logic below would be more or less the same), both on top of the same controller. That's an issue by itself.

>     Nordic's hardware QSPI takes care about this protocol.
>    I.e when you want to write some data to the flash you have to:

>   Check if memory is ready for transfer
    Send WREN (write enable) command
    Send write command with payload

>  You can do it using SPI, but you need to preapre 3 separate commands.
  In our implementation all steps are done by hardware, so you just have to tell where to write data          and peripherial will do the rest.

That is the main problem of your approach: never take a specific vendor hw implementation as a generalization. 
You'll end up creating an API that will be available only on that specific vendor hw.
(and again such hw optimization is not dual/quad/octal mode specific, I have seen it already on standard SPI controller). 
That said, if such hw optimization can be made available in a generic way and is a requirement, we'll be able to find a way that does not require a new API. Just an improvement of the existing SPI one. (I see already how we could achieve this)

In short, I still think we need to work on #17902. For instance, reading at what truly motivated this RFC, I can deduce that we need to expose the flash access hw optimization.

About DTS, as @pabigot mentioned it: we need to improve SPI configuration options exposure.
When it comes about io pins, it's not spi specific anyway: we do have this issue in all pinout based controllers. There was a attempt to fix it through a pinctrl DTS+API prototype. Anyway I am not going into this as DTS and pin muxing is another story by itself here. It is not related with the subject.

**Comment by kamlaz** [comment]
@tbursztyka 
Thank you for your input here - it is very precious.
In general aim of the QSPI is different than SPI. 
SPI is very generic, can work with many type of devices, usually using transceive.
QSPI was introduced as a protocol to communicate efficiently mainly with flash memories.
In SPI almost everything you have to do manually in contrast of QSPI, where hardware does almost everything for you.
I have look into other vendors QSPI implementation and most of them supports the same features:
	- hardware optimization of transfer (read/write/erase)
	- automatic pooling of flash memory (very useful feature)
As you mentioned 

> Only difference being that your use case can take advantage of dual/quad/octal modes (i.e.: more data lines for faster transaction) + your vendor specific hw optimization (see below)

These improvements brings time and processing power savings - that is important.

> That is the main problem of your approach: never take a specific vendor hw implementation as a generalization.

I can't agree, since other vendors QSPI implementation will fit to proposed one. Before I started composing QSPI API I spent some time to investigate how others have implemented their QSPI peripherial. In most cases they are composed of basic functions: read/write/erase/custom_cmd - the same as in my API.

At the end something very generic: 
In my opinion Zephyr should consists of :
	1. small target-oriented drivers
	2. drivers does elementary operation - are simple and reliable
And I am afraid that SPI driver will become spaghetti code that does everything.
So let SPI does SPI's things and QSPI what is QSPI designed for.

**Comment by pabigot** [comment]
> Note that a controller that can do dual/quad/octal spi, can still do the standard mode as well.
> So with you solution we would need 2 drivers: one SPI, one QSPI (knowing that 95% of the logic below would be more or less the same), both on top of the same controller.

@tbursztyka That's true in the general case (e.g. I think NXP's FlexSPI, and possibly STM32's Quad-SPI memory interface), but not for Nordic.  See the [Nordic QSPI peripheral overview](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fps_nrf52840%2Fqspi.html&cp=3_0_0_5_18).  It's a **very** specialized device.

@kamlaz Now that I think I understand a bit more about what other vendors do I believe you should implement a `nordic,qspi-nor` device that provides the Zephyr flash API on top of the Nordic QSPI peripheral.  If you find you need something beyond the existing API (e.g. XIP support, or custom command) you can add it through [device-specific extensions](https://docs.zephyrproject.org/latest/reference/drivers/index.html#device-specific-api-extensions).  Your devicetree specification is close to right for your use case, and it's *not* suitable for other vendors that use pinmux.

I don't find your argument for separating (Nordic) QSPI from flash compelling.  The [flash map](https://docs.zephyrproject.org/latest/reference/storage/flash_map/flash_map.html) is not part of the flash API: it's layered on top of the flash API.  The [existing flash API](https://docs.zephyrproject.org/latest/reference/peripherals/flash.html) is close to your proposed QSPI API, absent support for asynchronous operation (which could be added).

If you want to have a different API specifically for quad-spi memory interfaces you should coordinate with, at a minimum, STM32--but also the other vendors you say you've investigated--to confirm that your API is suitable for their Quad-SPI memory interface peripheral as well.  From experience I can say that will be a very long process.

My opinion, based on what I believe I know at this time.

**Comment by kamlaz** [comment]
@pabigot 

> you should implement a nordic,qspi-nor device that provides the Zephyr flash API on top of the Nordic QSPI peripheral.

I agree up to a point. Personally I do not think that creating `nordic,qspi-nor` has sense since:

-  it will be vendor-specific file. In my opinion flash-level API should not be **vendor specific**, but **generic** (i.e `jedec,qspi-nor`), just like `jedec,spi-nor`.
- completely different operational idea than `jedec,spi-nor`, 

In my opinion using `jedec,qspi-nor` has significant advantages:

-  it enables other vendors to use generic  `jedec,qspi-nor` API (and to plug in their own functions from their QSPI driver) 
-  provide operational consistency with `jedec,spi-nor`, where SPI driver is "below" flash-api

So the hierarchy will look like this:
image

**Comment by pabigot** [comment]
I'm getting confused because I can't distinguish APIs from drivers in your hierarchy, and because we're not rigorously distinguishing generic Single-/Dual-/Quad-/Octal-SPI from a "QSPI" that's specific to flash memories controlled by JEDEC commands.

Currently we have (1):
* a generic file-system API that's implemented by multiple drivers (FatFS, littlefs, ...).  This uses...
* a generic flash API that's implemented by multiple drivers (vendor-specific SOC flashes, and a generic JEDEC SPI-NOR driver).  The generic JEDEC driver uses...
* a generic (4-wire, single-) SPI API that's implemented by multiple drivers (Nordic SPI and SPIM, NXP LPSPI, ....).

I'm proposing you do (2):

* a generic file-system API that's implemented by multiple drivers (FatFS, littlefs, ...).  This uses...
* a generic flash API that's implemented by a custom driver for Nordic QSPI which supports only JEDEC flash memories.

For that we need one new driver implementation, with possibly (but not necessarily) some vendor-specific operations.  No new API, just a `nordic,qspi-nor` devicetree binding to provide enough information to create devices for the driver.  Simple, and you don't have to coordinate API design with everybody else.

I think the alternative you're proposing is (3):

* a generic file-system API that's implemented by multiple drivers (FatFS, littlefs, ...).  This uses...
* a generic flash API that's implemented by multiple drivers (all the drivers from (1) plus a new generic JEDEC Quad-SPI-NOR driver).  The generic JEDEC Quad-SPI driver uses...
* a generic Quad-SPI API that's implemented by multiple drivers (Nordic QSPI, and who else?).

You might be able to do this, but the first step is to figure out who else would want to implement this generic API.  NXP FlexSPI is a candidate.  Who else?

The next step is to demonstrate why the extension proposal in #17902 is not viable and a special Quad-SPI API is required.

If you can do that, you then have to come up with a vendor-independent API capable of doing Quad-SPI bus transactions.  What you've proposed here is far from that.  You need to get involvement from those vendors to ensure not only that they can implement it, but that it meets their customer needs and doesn't rule out extra capabilities that they might support, such as scatter/gather (iovec) transactions.

Beneath all that I remain deeply skeptical that the Nordic QSPI peripheral *can* operate as a generic Quad-SPI peripheral for something like an ADC where there is no 24/32-bit memory address involved in the transaction: I just don't see how you keep it from including the command address in the bus write/read transaction.  The strongest argument I can see for why #17902 doesn't work for Nordic would be that the QSPI peripheral can't be made to work with anything but a flash memory.  But at that point you have to justify why (2) is not the right path forward.

What's your goal here?  If it's to get fast serial memory support for Nordic customers into Zephyr then in my opinion the fastest path is (2): implement the Zephyr flash API directly.

**Comment by kamlaz** [comment]
> I think the alternative you're proposing is (3):
a generic file-system API that's implemented by multiple drivers (FatFS, littlefs, ...). This uses...
a generic flash API that's implemented by multiple drivers (all the drivers from (1) plus a new generic JEDEC Quad-SPI-NOR driver). The generic JEDEC Quad-SPI driver uses...
a generic Quad-SPI API that's implemented by multiple drivers (Nordic QSPI, and who else?).

That is **exactly** what I was thinking about!

> You might be able to do this, but the first step is to figure out who else would want to implement this generic API. NXP FlexSPI is a candidate. Who else?

To answer this question I had to look into few datasheets from other vendors. Please find the results below:
Most vendors has implemented QSPI in their chips. The main reason of implementing QSPI was managing communication with jedec flash devices:

- STM: AN4760, page 8

> The QUADSPI is a serial interface that allows the communication on four data lines between
a host (STM32) and an **external QSPI memory**

- NXP: AN4777, page 1

> Quad Serial Peripheral Interface (QuadSPI) is a
communications protocol used for communications between a
microcontroller and an **external flash memory**

- Microchip: Atmel-44065A, page 8

> The QSPI can be used to read data from the serial Flash memory allowing the CPU to execute code from
it (XIP execute in place). 

But Microchip extents this definition to:

> The QSPI can be used in SPI mode to interface to serial peripherals (such as ADCs, DACs, LCD
controllers, CAN controllers and sensors).

That is feature that could be considered, but has low importance in case of QSPI driver. It is due to fact that when QSPI is used in SPI mode, it uses only 4 lines (CS,SCK,MISO,MOSI) and two data lines are not used. So if anyone would like to communicate with i.e ADC that supports SPI, what is the aim of using QSPI instead of SPI? 
What is more, some vendors provides SPI mode in QSPI driver, but with limited functionality, i.e STM (AN4760, page 23):
image
This limitation makes from QSPI very poor SPI substitute. 

So in my opinion this kind of operation should not be supported, but proposed API enables user to do so (using `qspi_send_cmd`).
An finally answering your question : Most vendors should be interested in adding QSPI driver as jedec flash memory driver.

> The next step is to demonstrate why the extension proposal in #17902 is not viable and a special Quad-SPI API is required.

In my opinion:

- QSPI driver has to be specialized high performance driver for jedec-flash memories 
- Extending SPI API to QSPI functionality will create chaos, since one driver has to control two independent but not compatible peripherals (i.e DDR mode)
- Plugging QSPI API into generic JEDEC Quad-SPI driver will be a way more convenient and natural than SPI API.

**Comment by pabigot** [comment]
OK.  I won't press this further, but you need to get people from all those other vendors to agree that there's value to them to have a generic QSPI implementation, rather than just implement the flash API.  I'm not going to approve a new API unless there's clear evidence an existing one can't be made to work.

**Comment by nvlsianpu** [comment]
@MaureenHelm @erwango @franciscomunoz
As far I know some of NXP, STM and Microchip SoC incorporates QSPI peripheral.(**not the same as just QuadSPI**). Can you review this API introduction.
I think It would be great to have this as API or dev meeting meeting topic next week.

**Comment by hongshui3000** [comment]
The discussion here seems to be the octal-spi / q-spi flash memory controller API  rather than the general QSPI API ?

**Comment by nvlsianpu** [comment]
@hongshui3000 This is not really about implementation of 8/4 line SPI.
Author intention is to introduce API for the `Serial Memory Protocol Engine` with serial back-end using 4 line SPI. The peripheral `Serial Memory Protocol Engine` gives huge acceleration in I/O operation.
Wort to emphasize that such a peripheral not necessarily allow to RAW access to SPI bus at all.

**Comment by tbursztyka** [comment]
@kamlaz I will come with a proper RFC to fix #17902 , and in the mean time I'll try to address flash access optimization there. I still think it's possible to do it that way. However, if it proves to be not really possible (i.e. not addressing 100% of the issue as your API would) or crappy (creating awful addition to the SPI API and drivers) then yes we will go with a separate API.

File: include/drivers/qspi.h [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+
+/**
+ * @brief QSPI command structure
+ * Structure used for custom command usage.
+ *
+ * @param op_code is a command value (i.e 0x9F - get Jedec ID)
+ * @param tx_buf structure used for TX purposes. Can be NULL if not used.
+ * @param rx_buf structure used for RX purposes. Can be NULL if not used.
+ */
+struct qspi_cmd {
```
Comment (erwango) [inline include/drivers/qspi.h]: This struct should contain more information as more parameters might to be set per command:
-cmd: opcode, buswidth
-addr: nbytes, buswidth
-...

File: include/drivers/qspi.h [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+	 */
+	u8_t data_lines : 2;
+
+	/* Defines how many bits are used for address (8/16/24/32)
+	 * 0x00: 8 bit address field
+	 * 0x01: 16 bit address field
+	 * 0x02: 24 bit address field
+	 * 0x03: 32 bit address field
+	 */
+	u8_t address : 2;
```
Comment (erwango) [inline include/drivers/qspi.h]: data lines and address nbr of bytes can change per cmd.
These settings should be part of a more complete `qspi_cmd` structure

**Review by erwango — CHANGES_REQUESTED** [review]
From STM32 point of view this API would be easier to implement than the SPI extension on (#20564).
Though, API should be more flexible from CMD param point of view as some parameters may change per cmd rather than per session/memory. Then the `qspi_config` should only carry parameters that are static to a session and/or  memory specific (frequency, cs_pin , ...)

for CMD params, something like : https://elixir.bootlin.com/linux/v5.4/source/include/linux/spi/spi-mem.h would be ideal

File: include/drivers/qspi.h (hunk 2 above)
Comment (AutomatedBrewing) [reply include/drivers/qspi.h]: I can agree up to a point. In my opinion parameters like buswidth, address and others should be set by `configure` function. It is not without reason - when you want to change bus parameters, in general you have to uninit and then reinitialise perihperial. Proposed API contains function for that functionality - it is `configure`. 
In your proposition function `qspi_send_cmd` does the same as  `configure`. In my opinion each function has to do unique things.

Comment (kamlaz) [reply include/drivers/qspi.h]: I can agree up to a point. In my opinion parameters like buswidth, address and others should be set by `configure` function. It is not without reason - when you want to change bus parameters, in general you have to uninit and then reinitialise perihperial. Proposed API contains function for that functionality - it is `configure`.
In your proposition function `qspi_send_cmd `does the same as `configure`. In my opinion each function has to do unique things.

File: include/drivers/qspi.h (hunk 3 above)
Comment (kamlaz) [reply include/drivers/qspi.h]: Would you be so kind and give an example? I think that the only case when you have to change number of address bits is when you have two different memory types on one QSPI bus. In that case changing number of address bits is not a problem since you have to reconfigure peripheral anyway (to change used CS pin).
Please correct me if I am wrong.

Comment (erwango) [reply include/drivers/qspi.h]: Indeed when initializing the session with a memory, you need to perform configuration like specifying the number of lines you support and other things, but also, on a given memory, commands could have 1 to 4 bytes address length and be done one 1 to 4 lines.
On STM32 QSPI IP, this needs to be configured before sending of each command in order to avoid data corruption.

File: include/drivers/qspi.h (hunk 2 above)
Comment (erwango) [reply include/drivers/qspi.h]: > In your proposition function qspi_send_cmd does the same as configure. In my opinion each function has to do unique things.

No, I just have a different split of parameters. Some parameters you consider static for a given memory might actually change within a single transaction.

File: include/drivers/qspi.h (hunk 3 above)
Comment (kamlaz) [reply include/drivers/qspi.h]: Despite the fact that I understand your point of view, I still think that changing QSPI bus parameters should be done be `configure.` . Anyway, if we have to consider your case maybe this will do the job?
```
struct qspi_cmd {
	u8_t op_code;
	const struct qspi_buf *tx_buf;
	const struct qspi_buf *rx_buf;
        const struct qspi_cmd_conf  *conf;
};
```

where `qspi_cmd_conf` is
```
struct qspi_cmd_conf{
	u8_t data_lines : 2;
        u8_t address : 2;
};
```
But i do not think that this solution is good enough. I am curious what other interested think about it.

File: include/drivers/qspi.h (hunk 2 above)
Comment (kamlaz) [reply include/drivers/qspi.h]: Could you give an example? I have already work with some qspi memories and I have not found any situation to change `address  `size field (of course in case of single memory on the bus).
The only case when `data_lines` has to be reconfigured is when you are enabling Quad mode, but it is done only once.

Comment (erwango) [reply include/drivers/qspi.h]: I have the following example for a read on spi-nand memory (but which would require the exact same "spi memory" api), you need the following operations:
CMD PAGE_READ, nbytes 3,
CMD GET_FEATURE, nybytes 1,
CMD PAGE_READ_FROM_CACHE_OP, nbytes 2, 

Also note that for instance GET_FEATURE expects response on bus width of 1 line, even if memory has been configured in quad mode. Which justify the need to have the buswidth info at each command.

Comment (kamlaz) [reply include/drivers/qspi.h]: Thank you for the example! 
Indeed it shows situation when single data transfer is required. But it still do not touches the problem of changing the `address` field.
What do you think about this solution:
1. `qspi_write`/`qspi_read `commands operate with defined number of `data_lines` (i.e four)
2. `qspi_send_cmd` always operate with single data line.
That solution enables using fast four line transfer for read/write commands and still provide functionality for single line communication (i.e read status register).

File: include/drivers/qspi.h (hunk 3 above)
Comment (erwango) [reply include/drivers/qspi.h]: IIUC, Your proposal would work on QSPI up to 1-1-4, but from performance point of view this is sufficient.

Another point, what about dummy bytes handling?
Most read commands will use 1 dummy byte (fast read (0xB) for instance).

This would not address the OSPI needs, but maybe we can keep this for a future evolution.

**Comment by nzmichaelh** [comment]
FYI, I'm looking at adding QSPI support for the SAMD51 on the ItsyBitsy M4 Express.  This is used to talk to a on-board GD25B16C 2 MiB NOR flash.

On the SAMD51, the QSPI peripheral can run in "SPI mode" or "Serial Memory Mode".  In serial memory mode the flash is memory mapped and the peripheral automatically takes care of the command and address parts of the protocol.  It's easier to interface to and faster than SPI mode.

I'd like to see a 25xx JEDEC NOR flash driver that sits ontop of a QSPI driver similar to how spi_nor.c sits ontop of a SPI driver.  The QSPI driver would expose an API like the spi_nor_cmd_read / addr_read / cmd_write / addr_write currently used in spi_nor.c

File: include/drivers/qspi.h [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+	return api->configure(dev, config);
+}
+
+
+/**
+ * @brief Writes desired amount of data to the external flash memory.
+ *
+ * Note: This function is synchronous.
+ * Driver that realizes write functionality has to:
+ *  1. Check if memory is ready for operation - using RDSR
```
Comment (nzmichaelh) [inline include/drivers/qspi.h:299]: What's the advantage of requiring the driver do this instead of the JEDEC driver above it?

File: include/drivers/qspi.h [hunk 5]
```diff
@@ -0,0 +1,660 @@
...
+	return api->write(dev, tx_buf, address);
+}
+
+
+/**
+ * @brief Read desired amount of data from the external flash memory.
+ *
+ * Note: This function is synchronous.
+ * Driver that realizes read functionality has to:
+ *  1. Check if memory is ready for operation - using RDSR
```
Comment (nzmichaelh) [inline include/drivers/qspi.h:338]: Ditto - what's the advantage of having the driver do this?

File: include/drivers/qspi.h [hunk 6]
```diff
@@ -0,0 +1,660 @@
...
+ * For command issuing driver uses only one data line.
+ * This is universal function, so user can send
+ * i.e read/write/erase command using this command.
+ * If user would like to send command with address and
+ * payload (i.e write command), one must
+ * fill op_code(i.e PP - Page Program - 0x02)
+ * and tx_buff (address and data).
+ * For command structure refer to the memory manufacturers datasheet.
+ *
+ * Driver that realizes custom command functionality has to:
```
Comment (nzmichaelh) [inline include/drivers/qspi.h:386]: What does this mean?  Could you clarify?

File: include/drivers/qspi.h [hunk 7]
```diff
@@ -0,0 +1,660 @@
...
+				const struct qspi_cmd *cmd)
+{
+	const struct qspi_driver_api *api =
+		(const struct qspi_driver_api *)dev->driver_api;
+
+	return api->send_cmd(dev, cmd);
+}
+
+/**
+ * @brief Erases desired amount of flash memory.
```
Comment (nzmichaelh) [inline include/drivers/qspi.h:411]: This is a feature of JEDEC, not of QSPI.  Should this be in the JEDEC driver that calls this?

**Comment by pabigot** [comment]
> On the SAMD51, the QSPI peripheral can run in "SPI mode" or "Serial Memory Mode". In serial memory mode the flash is memory mapped and the peripheral automatically takes care of the command and address parts of the protocol. It's easier to interface to and faster than SPI mode.

@nzmichaelh Please see #25958.  In the current model you would implement the Zephyr flash API for the SAMD51 serial memory mode.

Since this PR has been moved to on-hold/abandoned, I'm going to close this for now: we should re-open it if it turns out we need a dedicated SPI interface.
